src/Avalonia.Styling/StyledElement.cs

```diff
@@ -404,7 +405,9 @@ void ISetLogicalParent.SetParent(ILogical? parent)
 
                 Parent = (IStyledElement?)parent;
 
-                if (_logicalRoot != null)
+                // We don't need to detach element from itself, if it implements ILogicalRoot, and it wasn't explicitly attached from OnAttachedToLogicalTreeCore.
+                if (_explicitlyAttachedToLogicalRoot
+                    && _logicalRoot != null)
```

Comment:
Here I want to replace boolean field with simple ```if (_logicalRoot != null && _logicalRoot != this)```. Or we also can get newRoot before detaching, and compare if it wasn't changed.
But I am not sure, can it break something else, @grokys?

Reply:
Apparently it already broke two tests.
I will review this PR later.